### DISCUSSION TITLE
是否可以用 Swift 或者 Rust 以這個框架寫輸入法？

### DISCUSSION THREAD

**Discussion by ShikiSuen** [fcitx/fcitx5#482] [Q&A] [answered]
C++ 編程門檻實在太高。
為了省事，我最近弄了基於 Swift 的天權星引擎：https://github.com/ShikiSuen/Megrez
只是我不知道 FCITX5 是否可以與 Swift 協作就是了。

**Comment by wengxt** [comment, marked as answer]
就像其他任何 C++ 和 Rust / Swift / Go 一樣，你需要把他們封裝成對應的 Library 給輸入法調用，用 C++ 寫個 glue 層。

例如
https://github.com/OpenBangla/OpenBangla-Keyboard/tree/develop/src/engine/fcitx
https://github.com/OpenBangla/riti/

一些題外話僅代表個人意見，Rust 門檻怕不是比如 C++ 更高。

**Reply by ShikiSuen** [reply]
Rust 只是隨口一問。我主要用 Swift。

**Reply by ShikiSuen** [reply]
顺便再问一下好了。也就是说，目前 FCITX5 的对外通讯方式仍旧是 C++ 而非 C？

------------

如果是的话，我这边输入法移植恐怕得等 Swift 5.7 完成 C++ Interop 才能开始了（至少目前是什么也做不了的）。

**Reply by wengxt** [reply]
几种选择，如上面说的，你可以让你的实现不出现任何 Foreign language -> C++ 的调用，只进行 C++ -> Foreign language 的调用。C++ -> foreign language 调用全走类 C 接口，go 或者 rust 都没有问题。因为对输入法引擎来说，其实实现是非常单纯的，干的事情基本上只有

1. 初始化一个全局对象存储数据
2. 初始化许多对象存储per client 的状态 （一些输入法库中 1 2 是一个整体，所有 client 使用同一个状态）
3. 把按键发送给 per client 的状态，并读取状态产生的数据返回给界面
4. 在一些情况下重置状态为空

必要和 C++ 的交互是很少的，不需要进行 foreign -> C++ 的调用，那些也就是写个 glue 层就好了

如果真的需要，那也就是给几个界面用到的一个类 C 接口就可以了

**Reply by ShikiSuen** [reply]
谢谢。那看来有点麻烦了。Swift 的 Structs 与 Enums 都不是 C++ 可以直接 -> 的东西。
要是有 macOS InputMethodKit 在 Linux 平台的对应兼容框架就好了。
